src/main/java/org/spongepowered/api/util/command/spec/CommandSpec.java

```diff
@@ -0,0 +1,392 @@
+     *
+     * @return This command's input tokenizer
+     */
+    public InputTokenizer getInputTokenizer() {
+        return this.argumentParser;
+    }
+
+
+    @Override
+    public Optional<CommandResult> process(CommandSource source, String arguments) throws CommandException {
+        checkPermission(source);
+        final CommandArgs args = new CommandArgs(arguments, getInputTokenizer().tokenize(arguments, false));
+        final CommandContext context = new CommandContext();
+        this.populateContext(source, args, context);
+        return Optional.of(getExecutor().execute(source, context));
+    }
+
+    @Override
+    public List<String> getSuggestions(CommandSource source, String arguments) throws CommandException {
+        CommandArgs args = new CommandArgs(arguments, getInputTokenizer().tokenize(arguments, true));
+        return complete(source, args, new CommandContext());
+    }
+
+    @Override
+    public boolean testPermission(CommandSource source) {
+        return this.permission == null || source.hasPermission(this.permission);
+    }
+
+    /**
+     * Get a short, one-line description used with this command if any is present
+     *
+     * @return the short description.
+     */
+    @Override
+    public Optional<Text> getShortDescription(CommandSource source) {
+        return Optional.fromNullable(this.description);
+    }
+
+    /**
+     * Get the usage for this command appropriate for the provided command source.
+     *
+     * @param source The source
+     * @return the usage for the source
+     */
+    @Override
+    public Text getUsage(CommandSource source) {
+        Preconditions.checkNotNull(source, "source");
+        return this.args.getUsage(source);
+    }
+
+    /**
+     * Return a longer description for this command. This description is composed of at least all present of the short description, the usage
+     * statement, and the extended description
+     * @param source The source to get the extended description for
+     * @return the extended description
+     */
+    @Override
+    public Optional<Text> getHelp(CommandSource source) {
+        Preconditions.checkNotNull(source, "source");
+        TextBuilder builder = Texts.builder();
+        Optional<Text> desc = getShortDescription(source);
+        if (desc.isPresent()) {
+            builder.append(desc.get(), CommandMessageFormatting.NEWLINE_TEXT);
+        }
+        builder.append(getUsage(source));
+        if (this.extendedDescription != null) {
+            builder.append(CommandMessageFormatting.NEWLINE_TEXT, this.extendedDescription);
+        }
+        return Optional.of(builder.build());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof CommandSpec)) {
+            return false;
+        }
+
+        CommandSpec that = (CommandSpec) o;
+
+        if (!this.args.equals(that.args)) {
```

Comment:
Same for Guava here as below, `Objects.equal()` will handle the null check for you:
```java
return Objects.equal(this.args, that.args)
        && Objects.equal(this.argumentParser, that.argumentParser)
        && ...;
```